File: packages/gbnf-experiment/data/20260909T011452Z_9f286468/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == "":
        return ["No input provided"]

    lines = src.split("\n")

    line_idx = 0
    # `lines[line_idx]` is falsy in JS both for an empty line and for running off the
    # end of the array, so an empty line halts the walk just like the end of input does.
    while line_idx < len(lines) and lines[line_idx] and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: List[str] = []
    for i in range(
        max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)), line_idx + 1
    ):
        lines_to_show.append(lines[i] if i < len(lines) else "undefined")
    return [
        *lines_to_show,
        " " * pos + "^",
    ]

File: packages/gbnf-experiment/data/20260909T125752Z_09b2c811/ported_implementation/src/utils/errors/build-error-position.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }
  const lines = src.split('\n');

  let lineIdx = 0;
  while (lines[lineIdx] && pos > lines[lineIdx].length - 1 && pos < src.length) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  for (
    let i = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
    i < lineIdx + 1;
    i++
  ) {
    linesToShow.push(lines[i]);
  }

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};
